src/managers/quorum.py

```diff
@@ -0,0 +1,194 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Manager for for handling quorum + ACL updates."""
+import logging
+import re
+import socket
+from typing import Set
+
+from charms.zookeeper.v0.client import (
+    MemberNotReadyError,
+    MembersSyncingError,
+    QuorumLeaderNotFoundError,
+    ZooKeeperManager,
+)
```

Comment:
**suggestion, out-of-scope** All of these now really look something we could put in the manager. Actually, `ZooKeeperManager` is EXACTLY the business logic to interact with the ZooKeeper server. I understand that right now this is in charm library, but these classes together with the QuorumManager really makes up the definition of the "managers" and they should probably stay in the same place

Reply:
The goal of the `client` charm library is not to contain business logic for the charm, but instead to function like an actual Python module, namely a wrapper for `kazoo` as a ZooKeeper service client.

It came about because I needed a way of issuing commands to ZooKeeper, from the Kafka charm. I'd like to keep it that way. The Kafka charms use it already for `broker_active`, and in future it might see use for other applications looking to connect to ZooKeeper.

I don't think it's a good idea to use a charm library to hide away core charm operational functionality, but it's **generic** and **useful** enough to justify it being in a shareable lib. I don't think any of these methods here meet either criteria.